### PR TITLE
Sign In and Sign Up implemented

### PR DESCRIPTION
**Pull request by FarazPourebrahim**

### CODE REVIEW

File: SET_internship/public/SignUpCard.png
Comment (atareversei) [inline SET_internship/public/SignUpCard.png:1]: use webp instead

File: SET_internship/src/App.css
Comment (atareversei) [inline SET_internship/src/App.css:1]: import standalone css files in this global css file to manage styles

File: SET_internship/src/App.tsx [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
     <>
-
+        <Router>
+            <Routes>
+                <Route path="/" element={<Navigate to="/sign_up" />} />
+                <Route path="/sign_up" element={<SignUp />} />
+                <Route path="/sign_in" element={<SignIn />} />
+                <Route path="/dashboard" element={<Dashboard />} />
+            </Routes>
+        </Router>
```
Comment (atareversei) [inline SET_internship/src/App.tsx]: extract theses routes into their own files
https://reactrouter.com/en/main/routers/create-browser-router

File: SET_internship/src/api/connect.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    PARSE = 'parse',
+    UNKNOWN = 'unknown',
+}
+
+type ApiResponse = {
+    success: boolean;
+    data?: any;
+    error?: string;
+    errorType?: ApiErrorType;
+}
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts]: discriminated union

File: SET_internship/src/api/connect.ts [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                    } else {
+                        result.success = false;
+                        result.error = 'Expected JSON response';
+                        result.errorType = ApiErrorType.PARSE;
+                    }
+                } catch {
+                    result.success = false;
+                    result.error = 'Failed to parse response';
+                    result.errorType = ApiErrorType.PARSE;
+                }
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts]: redundant try/catch block

File: SET_internship/src/api/endpoints.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+const BASE_URL: string = "http://45.159.149.130:99";
```
Comment (atareversei) [inline SET_internship/src/api/endpoints.ts]: add api version as a variable
add api endpoints

File: SET_internship/src/api/connect.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    constructor(baseURL: string) {
+        this.baseURL = baseURL;
+    }
+
+    private async request(endpoint: string, options?: RequestInit): Promise<ApiResponse> {
+        try {
+            const response = await fetch(`${this.baseURL}/${endpoint}`, options);
+
+            const result: ApiResponse = {
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts]: check res.status

File: SET_internship/src/api/connect.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+type ApiResponse = {
+    success: boolean;
+    data?: any;
+    error?: string;
+    errorType?: ApiErrorType;
+}
+
+class Connect {
+    private baseURL: string;
+
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts:35]: add static field for token

File: SET_internship/src/components/dashboard/Dashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+const Dashboard = () => {
+    const [activeElement, setActiveElement] = useState('NewTask');
+    const navigate = useNavigate();
+
+    useEffect(() => {
+
+            const connect : Connect = new Connect(BASE_URL);
+            async function handleConnection() {
+                const token = localStorage.getItem('authentication-token');
+                const data = await connect.get('api/v1/authenticated/me',{headers: {
```
Comment (atareversei) [inline SET_internship/src/components/dashboard/Dashboard.tsx]: refactor authorization header

File: SET_internship/src/components/signin/SignInForm.module.css [hunk 8]
```diff
@@ -0,0 +1,161 @@
+html, body {
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.module.css]:
```css
*, *::before, *::after {
 margin: 0;
 padding: 0;
 box-sizing: border-box;
}
```

File: SET_internship/src/components/signin/SignInForm.module.css [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    margin: 0;
+    padding: 0;
+    height: 100%;
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.module.css]:
```css
body {
 min-height: 100dvh;
}
```

File: SET_internship/src/components/signin/SignInForm.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        {
+            setPassword(value);
+        }
+    };
+
+    const handleSubmit = async (e) => {
+        e.preventDefault();
+
+        function requestBodyBuilder(identifier) {
+            const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.tsx]: take these into a common utility file or class

File: SET_internship/src/components/signin/SignInForm.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+            const phoneRegex = /^(0\d{10})$/;
+
+            let requestBody = {};
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.tsx]: create and api layer with strongly typed inputs and outputs

File: SET_internship/src/components/signin/SignInForm.tsx [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+                            className={styles["password-toggle"]}
+                        />
+                    </Form.Group>
+
+                    <Button type="submit" className={styles["btn-submit"]}>
+                        ورود
+                    </Button>
+
+                        <p className={styles["redirect-text"]}>
+                            <a onClick={() => navigate("/sign_up")} className={styles["link-spacing"]}>{PERSIAN.enter_with_otp}</a>
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.tsx]: https://reactrouter.com/en/main/components/link

File: SET_internship/src/utils/toast/toastStyles.css [hunk 13]
```diff
@@ -0,0 +1,18 @@
+*{
```
Comment (atareversei) [inline SET_internship/src/utils/toast/toastStyles.css]: :root

**Review by atareversei — CHANGES_REQUESTED** [review]

File: SET_internship/src/api/connect.ts [hunk 14]
```diff
@@ -1,115 +1,97 @@
 enum ApiErrorType {
     NETWORK = 'network',
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts]: unified enum types

File: SET_internship/src/api/connect.ts [hunk 15]
```diff
@@ -1,115 +1,97 @@
...
-    errorType?: ApiErrorType;
+type SuccessResponse<T> = {
+    success: true;
+    data: T;
 }
 
+type ErrorResponse = {
+    success: false;
+    error: string;
+    errorType: ApiErrorType;
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts:20]: form errors

File: SET_internship/src/api/connect.ts [hunk 16]
```diff
@@ -1,115 +1,99 @@
...
-                    result.errorType = ApiErrorType.PARSE;
-                }
-            
-
-
-            return result;
+            if (response.ok) {
+                return { success: true, data };
+            } else {
+                return {
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts:65]: check status

File: SET_internship/src/api/connect.ts [hunk 17]
```diff
@@ -1,115 +1,99 @@
...
-            }
-        );
+    public static put<T>(endpoint: string, body: any, options?: RequestInit): Promise<ApiResponse<T>> {
+        return this.request<T>(endpoint, {
+            method: 'PUT',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify(body),
+            ...options,
+        });
     }
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts:123]: add patch method

File: SET_internship/src/api/connect.ts [hunk 18]
```diff
@@ -1,115 +1,99 @@
...
-                    result.success = false;
-                    result.error = 'Failed to parse response';
-                    result.errorType = ApiErrorType.PARSE;
-                }
-            
-
-
-            return result;
+            if (response.ok) {
+                return { success: true, data };
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts]: data: data.data as T

File: SET_internship/src/api/api.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
+import Connect, {ApiResponse} from "./connect";
+import {endpoints} from "./endpoints";
+
+const API = {
+    async validateNumber<T> (phoneNumber: string): Promise<ApiResponse<T>> {
+        return await Connect.post(endpoints.validate_number, { 'mobile': phoneNumber });
+    },
```
Comment (atareversei) [inline SET_internship/src/api/api.ts]: remove generic type from function signature 
and define request and response types and add them to the function
and don't forget to pass the response type to generic argument slot

File: SET_internship/src/api/api.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+    async LogInWithPassword<T> (requestBody: any): Promise<ApiResponse<T>>{
+        return await Connect.post(endpoints.login_with_password, requestBody);
+    },
+
+    async getProfile<T> (): Promise<ApiResponse<T>> {
+        return await Connect.get('api/v1/authenticated/me', {
+            headers: {
+                'Authorization': `Bearer ${Connect.token}`,
```
Comment (atareversei) [inline SET_internship/src/api/api.ts]: move token assignment into connect class

File: SET_internship/src/api/endpoints.ts [hunk 21]
```diff
@@ -1,4 +1,8 @@
...
-export default BASE_URL;
\ No newline at end of file
+export const BASE_URL: string = "http://45.159.149.130:99";
+export const version: string = 'v1';
+export const endpoints = {
+    validate_number : `api/${version}/auth/user/validate`,
+    request_otp : `api/${version}/auth/user/request-otp`,
+    register : `api/${version}/auth/user/register`,
+    login_with_password : `api/${version}/auth/user/login-with-password`,
+}
```
Comment (atareversei) [inline SET_internship/src/api/endpoints.ts]: getter function

File: SET_internship/src/components/components/Anchor/Anchor.tsx [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+
+export function Anchor({
+                           color = "black",
+                           hover = "blue-underline",
+                           underline = "visible-underline",
+                           children,
+                           ...rest
+                       }: Props) {
+    return (
+        <Link className={`${s[color]} ${s[hover]} ${s[underline]}`} {...rest}>
```
Comment (atareversei) [inline SET_internship/src/components/components/Anchor/Anchor.tsx]: does this component handle navigating to absolute URLs like `https://google.com`?

File: SET_internship/src/components/components/Anchor/Anchor.tsx [hunk 23]
```diff
@@ -0,0 +1,29 @@
+import React from "react";
+import s from './Anchor.module.scss';
+import {Link, LinkProps} from "react-router-dom";
+
+type Underline = 'visible-underline' | 'disabled-underline';
+type Color = 'black' | 'white' | 'brown';
+type Hover = 'none' | 'underline' | 'blue-underline';
```
Comment (atareversei) [inline SET_internship/src/components/components/Anchor/Anchor.tsx]: extra type

File: SET_internship/src/context/UserContext.tsx [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+
+type UserContextType = {
+    user: any;
+    setUser: (value: any) => void;
+}
+
+const UserContext = createContext<UserContextType | undefined>(undefined);
+
+export const UserProvider= ({ children }) => {
+    const [user, setUser] = useState<any>(null);
```
Comment (atareversei) [inline SET_internship/src/context/UserContext.tsx:14]: add use effect

File: SET_internship/src/components/signin/SignInForm.module.css [hunk 25]
```diff
@@ -1,8 +1,11 @@
...
+*, *::before, *::after {
     margin: 0;
     padding: 0;
-    height: 100%;
-    overflow: hidden;
+    box-sizing: border-box;
+}
+
+body {
+    min-height: 100vh;
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.module.css]: dvh

File: SET_internship/src/components/signin/SignInForm.tsx [hunk 26]
```diff
@@ -56,38 +72,40 @@ const SignInForm = () => {
             return requestBody;
         }
 
-        const connect : Connect = new Connect(BASE_URL);
+        Connect.configure(BASE_URL);
```
Comment (atareversei) [inline SET_internship/src/components/signin/SignInForm.tsx]: add config before app has been initialized

**Review by atareversei — CHANGES_REQUESTED** [review]

File: SET_internship/src/components/components/Anchor/Anchor.tsx (hunk 22 above)
Comment (FarazPourebrahim) [reply SET_internship/src/components/components/Anchor/Anchor.tsx]: Yes it does.

File: SET_internship/src/api/api.ts (hunk 19 above)
Comment (FarazPourebrahim) [reply SET_internship/src/api/api.ts]: I have a problem with this

File: SET_internship/src/api/api.ts
Comment (atareversei) [inline SET_internship/src/api/api.ts:1]: move this file into a new directory call `/services` and rename this file to `auth.ts`

File: SET_internship/src/api/connect.ts [hunk 27]
```diff
@@ -1,44 +1,70 @@
...
     data: T;
+    message: string;
+    status : boolean;
 }
 
 type ErrorResponse = {
     success: false;
     error: string;
     errorType: ApiErrorType;
+    formError?: {};
```
Comment (atareversei) [inline SET_internship/src/api/connect.ts:21]: `Record<string, string[]>` or `{ [key: string]: string[] }`

File: SET_internship/src/components/signup/SignUpForm.tsx [hunk 28]
```diff
@@ -57,33 +56,31 @@ const SignUpForm = () => {
...
                     }
 
                 } else if (phase === Phases.OTP) {
-                    const data = await API.register(phoneNumber,otpCode);
-                    if (data.success) {
-                        localStorage.setItem('authentication-token', data.data.data.token);
-                        Connect.token = data.data.data.token;
+                    const json = await API.register({mobile:phoneNumber, otp:otpCode});
+                    if (json.success) {
+                        localStorage.setItem('authentication-token', json.data.token);
```
Comment (atareversei) [inline SET_internship/src/components/signup/SignUpForm.tsx:78]: convert to a function on context

File: SET_internship/src/components/signup/SignUpForm.tsx [hunk 29]
```diff
@@ -149,7 +146,7 @@ const SignUpForm = () => {
                             {loading ? <Spinner animation="border" size="sm" /> : PERSIAN.sign_in_or_up}
                         </Button>
                         <p className={styles["redirect-text"]}>
-                            <Anchor color={'black'} hover={"blue-underline"} underline={"disabled-underline"} onClick={() => setPhase(Phases.PHONE_NUMBER)} to={'/sign_up'}>
+                            <Anchor color={'black'} underline={"on-hover"} onClick={() => setPhase(Phases.PHONE_NUMBER)} to={'/sign_up'}>
```
Comment (atareversei) [inline SET_internship/src/components/signup/SignUpForm.tsx:149]: redundant setPhase

**Review by atareversei — CHANGES_REQUESTED** [review]

**Pull request merged**